atc/db/team.go

```diff
@@ -1286,6 +1296,47 @@ func saveResource(tx Tx, resource atc.ResourceConfig, pipelineID int) (int, erro
 	return resourceID, nil
 }
 
+func existingResourceConfig(tx Tx, name string, pipelineID int) (atc.ResourceConfig, string, error) {
+	var (
+		configBlob, nonce sql.NullString
+		oldType           string
+	)
+	row := psql.Select("config", "type", "nonce").
+		From("resources").
+		Where(sq.Eq{
+			"name":        name,
+			"pipeline_id": pipelineID,
+		}).
+		RunWith(tx).
+		QueryRow()
+	err := row.Scan(&configBlob, &oldType, &nonce)
+	if err != nil && err != sql.ErrNoRows {
+		if err == sql.ErrNoRows {
+			return atc.ResourceConfig{}, "", nil
+		}
+		return atc.ResourceConfig{}, "", err
+	}
+
+	var noncense *string
+	if nonce.Valid {
+		noncense = &nonce.String
+	}
+
+	oldConfig := atc.ResourceConfig{}
+	if configBlob.Valid {
+		decryptedConfig, err := tx.EncryptionStrategy().Decrypt(configBlob.String, noncense)
```

Comment:
Hmm would it be possible for us to compare the encrypted payloads against each other to avoid needing to decrypt every config for every resource? Doing that will also cause there to be a diff if the db encryption key is rotated so maybe after we compare the encrypted payloads to be different, we could then check for a diff in the underlying config old vs new? That being said I'm not sure 100% if we can just compare the encrypted configs haha

Reply:
This would be a nice-to-have I think - it would require splitting out the fields in the encrypted blob that _are_ allowed to change (icon etc.) from the source which would be a much bigger change. I don't disagree that it would be nice but I don't want to scope creep the fixing of this bug too much for a performance improvement

Reply:
I could be missing something but I don't think I completely understand why it will require splitting out the fields in the encrypted blob? Would we not be able to straight up compare the old vs new encrypted payloads and most of the time they will be the same because nothing has changed. But if there is a diff then we decrypt it and do the same comparisons that we do right now (comparing the source and the type)?

Reply:
The encrypted blob is the whole config - not just the source. We don't necessarily want to reset the versions if e.g. the icon (which is part of the config) has changed. If we just compared the encrypted bit it would change if the icon changed I think

Reply:
Yep that works - it is a bit of added complexity but should improve the throughput as _most_ resources probably don't change on your average SavePipeline call 👍